### ISSUE TITLE
Column Naming in join is Duplicated when Due to First Suffix but not when Due to Second Suffix

### ISSUE THREAD

**Issue by billdenney** [tidyverse/dplyr#3266]
This is partly related to #3243.

When joining by specified column names, if names will duplicate due to the first suffix, then duplicated names are generated.  If names will duplicated from the second suffix, names are specific.  I would expect the latter behavior (the second pair of examples) and not the former (the first pair of examples).

``` r
library(dplyr, quietly = TRUE)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
d1 <- data.frame(A = 1, A.x = 2)
d2 <- data.frame(B = 3, A.x = 4, A = 5)
full_join(d1, d2, by = "A.x")
#>   A.x A.x  B A.y
#> 1   1   2 NA  NA
#> 2  NA   4  3   5
full_join(d2, d1, by = "A.x")
#>    B A.x A.x.x A.y
#> 1  3   4     5  NA
#> 2 NA   2    NA   1

d1 <- data.frame(A = 1, A.y = 2)
d2 <- data.frame(B = 3, A.y = 4, A = 5)
full_join(d1, d2, by = "A.y")
#>   A.x A.y  B A.y.y
#> 1   1   2 NA    NA
#> 2  NA   4  3     5
full_join(d2, d1, by = "A.y")
#>    B A.y A.x A.y.y
#> 1  3   4   5    NA
#> 2 NA   2  NA     1
```

**Comment by krlmlr** [comment]
@billdenney: The bugfix causes a more severe problem, #3307, which leads to downstream failures for at least three packages. I might have to revert it and fix resolution of name clashes differently (#3425) in a major release. What's your use case?

**Comment by billdenney** [comment]
My use case was related to a dataset received from a client where they had many variants of the same column name (a long test with one column per response), and I had to merge slightly different variants of these where the same person responded.

It was a light use case where in my particular example, I just went about it another way (gather first then merge).  I just found the issue then.

This bug was not a high priority for me, but I will probably experience it again in the future were the change reverted without an equivalent fix.

The short of it is: if you need to revert this for a while and replace it another way later, that's ok with me. (And, thanks for letting me know the reversion is coming.)

**Comment by krlmlr** [comment]
Will be resolved as part of #3425.
